### PR TITLE
 Disallow nullable suppression in patterns

### PR DESCRIPTION
**Pull request by alrz**
Fixes https://github.com/dotnet/roslyn/issues/54071

### CODE REVIEW

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

**Comment by alrz** [comment]
Looks like suppression operator *is* allowed but only in *some* cases, presumably only a coincidence because in those other cases parser couldn't handle it, not a explicit check. How should we approach this?

**Comment by jcouv** [comment]
> Looks like suppression operator is allowed but only in some cases [...]How should we approach this?

Given that a suppression in such place is nonsense, I think we can just disallow and treat the old behavior as a bug. We'd document the (minor) break just in case.

**Comment by alrz** [comment]
I've seen `is null!` usages but it's just wrong and misleading. I suppose we need compat council sign off on this. 
Where should we document the breaking change, post VS2019? And what version we're targeting for the fix?

**Comment by jcouv** [comment]
Will start email thread with compat council.
This would be documented in a new file for DotNet 6, similar to this one from DotNet 5: https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20DotNet%205.md

**Comment by jcouv** [comment]
Sorry for the delay. Before I send the email out, let's clarify what is the proposed scope of break. Is it just on number literals and `null`?

**Comment by alrz** [comment]
All literal expressions can be suppressed right now, wherever they appear in a constant pattern. Only identifiers gave syntax errors previously because we couldn't parse `!` after a type-like. 

Note this PR doesn't cover the case where a subexpression is suppressed, we may need to descend into the expression more deeply.

**Comment by alrz** [comment]
>Note this PR doesn't cover the case where a subexpression is suppressed, we may need to descend into the expression more deeply.

This would probably have even less impact than the top-level case but I'm not sure if we want it. 

I think we could flag the pattern binder and report on any suppression instead of walking down the whole thing *again* here.

**Comment by jcouv** [comment]
> case where a subexpression is suppressed

@alrz I'm not sure where we stand on subexpressions. Could we start with a couple of examples to clarify the discussion?

**Comment by alrz** [comment]
Both of these checks are not covered for subexpressions.

```cs
public class C {
    bool b;
    public void M() {
        
        if (this is { b: default }) {} // error
        if (this is { b: true! }) {} // proposed to warn
        
        if (this is { b: 0 == default }) {} // no error
        if (this is { b: 1 == 1! }) {} // no warning

    }
}
```

**Comment by jcouv** [comment]
I think it's fine to leave those last two line as "no error". Let's stick with the PR as scoped :-)

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review. Thanks

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review. Thanks

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:230]: I feel like the particular control flow here is a little weird. i.e. this could be `return e` and the `continue`s could be `break`s. However this seems subjective enough that I won't block the PR about it.

**Review by RikkiGibson — APPROVED** [review]

**Comment by jcouv** [comment]
CI integration legs won't pass until https://github.com/dotnet/roslyn/pull/55681 is merged.

**Pull request merged**
